sat/apiclient/fas.py

```diff
@@ -24,14 +24,14 @@
 """
 Client for querying the Firmware Action Service (FAS) API.
 """
+from datetime import datetime, timedelta
 import json
 import logging
 import time
-from datetime import datetime, timedelta
 
+from csm_api_client.service.gateway import APIError, APIGatewayClient
```

Comment:
There appears to be a type-checking error worth looking into on line 350 of this file. Here's the relevant snippet of code:

```python
        # create a system snapshot set to expire in 10 minutes
        payload = {'name': name, 'expirationTime': later}

        # filter on xnames if provided.
        if xnames:
            payload['stateComponentFilter'] = {'xnames': xnames}

        payload = json.dumps(payload, cls=_DateTimeEncoder)

        try:
            self.post('snapshots', payload=payload)
        except APIError as err:
            raise APIError('Error when posting new snapshot: {}'.format(err))
```

The error I'm seeing in PyCharm says, "Expected type 'Optional[dict]', got 'str' instead". Can you look into this and see if it's going to be a problem? Perhaps test `sat firmware` to double check.

Reply:
This is partly an issue with how this was originally written and how I changed the type annotations on `APIGatewayClient.post()` in `csm-api-client`. The `payload` argument is passed to the `data` argument of  `requests.request()`, which according to the [requests docs](https://requests.readthedocs.io/en/latest/api/#requests.request) can be:
> Dictionary, list of tuples, bytes, or file-like object to send in the body of the [Request](https://requests.readthedocs.io/en/latest/api/#requests.Request).

This corresponds to the type signature `Optional[Union[Dict, List[Tuple[...]], bytes, IO]]`. However, `APIGatewayClient.post()` only accepts `Optional[Dict]`. This restriction is not strictly necessary, but it corresponds to how it is being used already in SAT. It is fortunately fairly easy to work around in the caller by performing type conversions on the dict values before passing in the dict to the `post()` method.